src/idpf.rs

```diff
@@ -0,0 +1,1852 @@
+    P: Prg<L>,
+{
+    let bits = input.len();
+
+    let initial_seeds: [Seed<L>; 2] = [
+        Seed::from_rand_source(rand_source)?,
+        Seed::from_rand_source(rand_source)?,
+    ];
+
+    let mut seeds = [initial_seeds[0].0, initial_seeds[1].0];
+    let mut control_bits = [Choice::from(0u8), Choice::from(1u8)];
+    let mut inner_correction_words = Vec::with_capacity(bits - 1);
+
+    for (level, value) in inner_values.into_iter().enumerate() {
+        if level >= bits - 1 {
+            return Err(
+                IdpfError::InvalidParameter("too many values were supplied".to_string()).into(),
+            );
+        }
+        inner_correction_words.push(generate_correction_word::<FI, P, L, OUT_LEN>(
+            Choice::from(input[level] as u8),
+            value,
+            &mut seeds,
+            &mut control_bits,
+        ));
+    }
+    if inner_correction_words.len() != bits - 1 {
+        return Err(IdpfError::InvalidParameter("too few values were supplied".to_string()).into());
+    }
+    let leaf_correction_word = generate_correction_word::<FL, P, L, OUT_LEN>(
+        Choice::from(input[bits - 1] as u8),
+        leaf_value,
+        &mut seeds,
+        &mut control_bits,
+    );
+    let public_share = IdpfPoplarPublicShare {
+        inner_correction_words,
+        leaf_correction_word,
+    };
+
+    Ok((public_share, initial_seeds))
+}
+
+/// The IdpfPoplar key generation algorithm.
+///
+/// Generate and return a sequence of IDPF shares for `input`. The parameters `inner_values`
+/// and `leaf_value` provide the output values for each successive level of the prefix tree.
+pub fn gen<M, P, const L: usize, const OUT_LEN: usize>(
+    input: &IdpfInput,
+    inner_values: M,
+    leaf_value: [Field255; OUT_LEN],
+) -> Result<
+    (
+        IdpfPoplarPublicShare<Field64, Field255, L, OUT_LEN>,
+        [Seed<L>; 2],
+    ),
+    VdafError,
+>
+where
+    M: IntoIterator<Item = [Field64; OUT_LEN]>,
+    P: Prg<L>,
+{
+    if input.is_empty() {
+        return Err(IdpfError::InvalidParameter("invalid number of bits: 0".to_string()).into());
+    }
+    gen_with_rand_source::<_, _, _, P, L, OUT_LEN>(
+        input,
+        inner_values,
+        leaf_value,
+        getrandom::getrandom,
+    )
+}
+
+/// Evaluate an IDPF share on `prefix`, starting from a particular tree level with known
+/// intermediate values.
+#[allow(clippy::too_many_arguments)]
+fn eval_inner<P, const L: usize, const OUT_LEN: usize>(
+    bits: usize,
+    which_aggregator: bool,
+    public_share: &IdpfPoplarPublicShare<Field64, Field255, L, OUT_LEN>,
+    start_level: usize,
+    mut seed: [u8; L],
+    mut control_bit: Choice,
+    prefix: &IdpfInput,
+    cache: &mut dyn IdpfCache<L>,
+) -> Result<IdpfOutputShare<OUT_LEN, Field64, Field255>, IdpfError>
+where
+    P: Prg<L>,
+{
+    let mut last_inner_output = None;
+    for ((correction_word, input_bit), level) in public_share.inner_correction_words[start_level..]
```

Comment:
Hmm, something looks fishy here. Suppose we use `NoCache`. Do we end up evaluating a given prefix more than once?

```
eval():
   for start_level in 0..bits:
      eval_inner(start_level)

eval_inner(start_level):
   for level in start_level..bits:
       eval_next()
```

Reply:
eval() only calls eval_inner() once, and then breaks out of its loop, so we're only doing evaluation at each node once.

Reply:
Comments cleared up my confusion, thanks